# Feature Selection with Filtering Method | Constant, Quasi Constant and Duplicate Feature Removal

Unnecessary and redundant features not only slow down the training time of an algorithm, but they also affect the performance of the algorithm.

### What is filter method?

In [40]:
# Importing required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.feature_selection import VarianceThreshold

In [60]:
# load the dataset
data = pd.read_csv(r"C:\Users\Acer\Downloads\PYTHON\feature_selection\Data-Files-for-Feature-Selection-master\santander-train.csv", nrows = 20000)
data.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  \
0                        0                        0                0.0   
1                        0                        0                0.0   
2                        0                        0                0.0   
3                        0                        0                0.0   
4                        0                        0                0.0   

   imp_op_var41_comer_ult1  imp_op_var41_comer_ult3  imp_op_var41_efect_ult1  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                    195.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var41_efect_ult3  imp_op_var41_ult1  imp_op_var39_efect_ult1  \
0                      0.0                0.0                      0.0   
1                      0.0                0.0                      0.0   
2                      0.0                0.0                      0.0   
3                      0.0              195.0                      0.0   
4                      0.0                0.0                      0.0   

   imp_op_var39_efect_ult3  imp_op_var39_ult1  imp_sal_var16_ult1  ind_var1_0  \
0                      0.0                0.0                 0.0           0   
1                      0.0                0.0                 0.0           0   
2                      0.0                0.0                 0.0           0   
3                      0.0              195.0                 0.0           0   
4                      0.0                0.0                 0.0           0   

   ind_var1  ind_var2_0  ind_var2  ind_var5_0  ind_var5  ind_var6_0  ind_var6  \
0         0           0         0           1         0           0         0   
1         0           0         0           1         0           0         0   
2         0           0         0           1         1           0         0   
3         0           0         0           1         1           0         0   
4         0           0         0           1         0           0         0   

   ind_var8_0  ind_var8  ind_var12_0  ind_var12  ind_var13_0  \
0           0         0            0          0            0   
1           0         0            0          0            1   
2           0         0            0          0            0   
3           0         0            0          0            0   
4           0         0            1          1            0   

   ind_var13_corto_0  ind_var13_corto  ind_var13_largo_0  ind_var13_largo  \
0                  0                0                  0                0   
1                  1                1                  0                0   
2                  0                0                  0                0   
3                  0                0                  0                0   
4   

In [43]:
data.shape

(20000, 371)

In [50]:
# To show the all columns
pd.set_option("display.max_columns", 500)
pd.set_option("display.max_rows", 20000)

In [51]:
#checing for missing values
data.isna().sum()

ID                               0
var3                             0
var15                            0
imp_ent_var16_ult1               0
imp_op_var39_comer_ult1          0
imp_op_var39_comer_ult3          0
imp_op_var40_comer_ult1          0
imp_op_var40_comer_ult3          0
imp_op_var40_efect_ult1          0
imp_op_var40_efect_ult3          0
imp_op_var40_ult1                0
imp_op_var41_comer_ult1          0
imp_op_var41_comer_ult3          0
imp_op_var41_efect_ult1          0
imp_op_var41_efect_ult3          0
imp_op_var41_ult1                0
imp_op_var39_efect_ult1          0
imp_op_var39_efect_ult3          0
imp_op_var39_ult1                0
imp_sal_var16_ult1               0
ind_var1_0                       0
ind_var1                         0
ind_var2_0                       0
ind_var2                         0
ind_var5_0                       0
ind_var5                         0
ind_var6_0                       0
ind_var6                         0
ind_var8_0          

In [52]:
X = data.drop('TARGET', axis = 1)
y = data['TARGET']
X.shape, y.shape

((20000, 370), (20000,))

In [53]:
# train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0, stratify =y) 

# Constant Features Removal
In this, first we need to create a variance threshold and then fit the model with training set of the data.

In [54]:
constant_filter = VarianceThreshold(threshold=0)
constant_filter.fit(X_train)

VarianceThreshold(threshold=0)

Let’s get the number of features left after removing constant features.

In [58]:
constant_filter.get_support().sum()

291

 constant features list.

In [56]:
constant_list = [not temp for temp in constant_filter.get_support()]
constant_list

[False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 Fa

the name of features which are constant.



In [59]:
X.columns[constant_list]

Index(['ind_var2_0', 'ind_var2', 'ind_var13_medio_0', 'ind_var13_medio',
       'ind_var18_0', 'ind_var18', 'ind_var27_0', 'ind_var28_0', 'ind_var28',
       'ind_var27', 'ind_var34_0', 'ind_var34', 'ind_var41', 'ind_var46_0',
       'ind_var46', 'num_var13_medio_0', 'num_var13_medio', 'num_var18_0',
       'num_var18', 'num_var27_0', 'num_var28_0', 'num_var28', 'num_var27',
       'num_var34_0', 'num_var34', 'num_var41', 'num_var46_0', 'num_var46',
       'saldo_var13_medio', 'saldo_var18', 'saldo_var28', 'saldo_var27',
       'saldo_var34', 'saldo_var41', 'saldo_var46',
       'delta_imp_amort_var18_1y3', 'delta_imp_amort_var34_1y3',
       'delta_imp_reemb_var33_1y3', 'delta_imp_trasp_var17_out_1y3',
       'delta_imp_trasp_var33_out_1y3', 'delta_num_reemb_var33_1y3',
       'delta_num_trasp_var17_out_1y3', 'delta_num_trasp_var33_out_1y3',
       'imp_amort_var18_hace3', 'imp_amort_var18_ult1',
       'imp_amort_var34_hace3', 'imp_amort_var34_ult1', 'imp_var7_emit_ult1',
       'imp

In [72]:
X[['ind_var2_0', 'ind_var2', 'ind_var13_medio_0', 'ind_var13_medio',
       'ind_var18_0', 'ind_var18', 'ind_var27_0', 'ind_var28_0', 'ind_var28',
       'ind_var27', 'ind_var34_0', 'ind_var34', 'ind_var41', 'ind_var46_0',
       'ind_var46', 'num_var13_medio_0', 'num_var13_medio', 'num_var18_0',
       'num_var18', 'num_var27_0', 'num_var28_0', 'num_var28', 'num_var27',
       'num_var34_0', 'num_var34', 'num_var41', 'num_var46_0', 'num_var46',
       'saldo_var13_medio', 'saldo_var18', 'saldo_var28', 'saldo_var27',
       'saldo_var34', 'saldo_var41', 'saldo_var46',
       'delta_imp_amort_var18_1y3', 'delta_imp_amort_var34_1y3',
       'delta_imp_reemb_var33_1y3', 'delta_imp_trasp_var17_out_1y3',
       'delta_imp_trasp_var33_out_1y3', 'delta_num_reemb_var33_1y3',
       'delta_num_trasp_var17_out_1y3', 'delta_num_trasp_var33_out_1y3',
       'imp_amort_var18_hace3', 'imp_amort_var18_ult1',
       'imp_amort_var34_hace3', 'imp_amort_var34_ult1', 'imp_var7_emit_ult1',
       'imp_reemb_var13_hace3', 'imp_reemb_var17_hace3',
       'imp_reemb_var33_hace3', 'imp_reemb_var33_ult1',
       'imp_trasp_var17_in_hace3', 'imp_trasp_var17_out_hace3',
       'imp_trasp_var17_out_ult1', 'imp_trasp_var33_in_hace3',
       'imp_trasp_var33_out_hace3', 'imp_trasp_var33_out_ult1',
       'ind_var7_emit_ult1', 'num_var2_0_ult1', 'num_var2_ult1',
       'num_var7_emit_ult1', 'num_meses_var13_medio_ult3',
       'num_reemb_var13_hace3', 'num_reemb_var17_hace3',
       'num_reemb_var33_hace3', 'num_reemb_var33_ult1',
       'num_trasp_var17_in_hace3', 'num_trasp_var17_out_hace3',
       'num_trasp_var17_out_ult1', 'num_trasp_var33_in_hace3',
       'num_trasp_var33_out_hace3', 'num_trasp_var33_out_ult1',
       'saldo_var2_ult1', 'saldo_medio_var13_medio_hace2',
       'saldo_medio_var13_medio_hace3', 'saldo_medio_var13_medio_ult1',
       'saldo_medio_var13_medio_ult3', 'saldo_medio_var29_hace3']].head(20)

ind_var2_0  ind_var2  ind_var13_medio_0  ind_var13_medio  ind_var18_0  \
0            0         0                  0                0            0   
1            0         0                  0                0            0   
2            0         0                  0                0            0   
3            0         0                  0                0            0   
4            0         0                  0                0            0   
5            0         0                  0                0            0   
6            0         0                  0                0            0   
7            0         0                  0                0            0   
8            0         0                  0                0            0   
9            0         0                  0                0            0   
10           0         0                  0                0            0   
11           0         0                  0                0            0   
12           0         0                  0                0            0   
13           0         0                  0                0            0   
14           0         0                  0                0            0   
15           0         0                  0                0            0   
16           0         0                  0                0            0   
17           0         0                  0                0            0   
18           0         0                  0                0            0   
19           0         0                  0                0            0   

    ind_var18  ind_var27_0  ind_var28_0  ind_var28  ind_var27  ind_var34_0  \
0           0            0            0          0          0            0   
1           0            0            0          0          0            0   
2           0            0            0          0          0            0   
3           0            0            0          0          0            0   
4           0            0            0          0          0            0   
5           0            0            0          0          0            0   
6           0            0            0          0          0            0   
7           0            0            0          0          0            0   
8           0            0            0          0          0            0   
9           0            0            0          0          0            0   
10          0            0            0          0          0            0   
11          0            0            0          0          0            0   
12          0            0            0          0          0            0   
13          0            0            0          0          0            0   
14          0            0            0          0          0            0   
15          0            0            0          0          0            0   
16          0            0            0          0          0            0   
17          0            0            0          0          0            0   
18          0            0            0          0          0            0   
19          0            0            0          0          0            0   

    ind_var34  ind_var41  ind_var46_0  ind_var46  num_var13_medio_0  \
0           0          0            0          0                  0   
1           0          0            0          0                  0   
2           0          0            0          0                  0   
3           0          0            0          0                  0   
4           0          0            0          0                  0   
5           0          0            0          0                  0   
6           0          0            0          0                  0   
7           0          0            0          0                  0   
8           0          0            0          0                  0   
9           0          0            0

 transform the x_train and x_test datasets into non constant datasets.

In [66]:
X_train_filter = constant_filter.transform(X_train)
X_test_filter = constant_filter.transform(X_test)

In [73]:
X_train_filter.shape, X_test_filter.shape, X_train.shape

((16000, 291), (4000, 291), (16000, 370))

## Quasi constant feature removal

In [74]:
quasi_constant_filter = VarianceThreshold(threshold=0.01)
quasi_constant_filter.fit(X_train_filter)

VarianceThreshold(threshold=0.01)

TO see how many features are non quasi constant:

In [75]:
quasi_constant_filter.get_support().sum()

245

To remove those quasi constant features, we need to apply transform on quasi transform filter object.

In [76]:
X_train_quasi_filter = quasi_constant_filter.transform(X_train_filter)
X_test_quasi_filter = quasi_constant_filter.transform(X_test_filter)
X_train_quasi_filter.shape, X_test_quasi_filter.shape

((16000, 245), (4000, 245))

In [77]:
X_train_T = X_train_quasi_filter.T
X_test_T = X_test_quasi_filter.T
type(X_train_T)

numpy.ndarray

In [78]:
#Let’s change into pandas dataframe.
X_train_T = pd.DataFrame(X_train_T)
X_test_T = pd.DataFrame(X_test_T)

In [79]:
#To get the duplicate features.
X_train_T.duplicated().sum()

18

In [80]:
duplicated_features = X_train_T.duplicated()
duplicated_features

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28     False
29     False
30     False
31     False
32     False
33     False
34     False
35     False
36     False
37      True
38     False
39      True
40     False
41     False
42     False
43      True
44     False
45      True
46     False
47     False
48     False
49     False
50     False
51     False
52     False
53     False
54     False
55     False
56     False
57     False
58     False
59     False
60     False
61     False
62     False
63     False
64     False
65     False
66     False
67     False
68     False
69     False
70     False
71      True
72     False
73      True
74     False
75     False
76     False

In [81]:
features_to_keep = [not index for index in duplicated_features]

In [82]:
#Let’s do transpose again to get the original shape.

X_train_unique = X_train_T[features_to_keep].T
X_test_unique = X_test_T[features_to_keep].T

In [83]:
X_train_unique.shape, X_train.shape

((16000, 227), (16000, 370))

#### Here, we can observe original dataset has 370 features and after removal of quasi constant, constant and duplicate features we have 227 features.

## Build ML model and compare the performance of the selected feature
Let’s go ahead and compare the model between original dataset and transformed dataset.

Here we are going to build random forest classifier.

In [84]:
def run_randomForest(X_train, X_test, y_train, y_test):
    clf = RandomForestClassifier(n_estimators=100, random_state=0, n_jobs=-1)
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    print('Accuracy on test set: ')
    print(accuracy_score(y_test, y_pred))

In [85]:
%%time
run_randomForest(X_train_unique, X_test_unique, y_train, y_test) #accuracy with respect to selected features

Accuracy on test set: 
0.95875
Wall time: 985 ms


In [86]:
%%time
run_randomForest(X_train, X_test, y_train, y_test)    #accuracy with respect to origina dataset

Accuracy on test set: 
0.9585
Wall time: 1.17 s
